module/zfs/dmu_tx.c

```diff
@@ -154,17 +147,37 @@ dmu_tx_hold_object_impl(dmu_tx_t *tx, objset_t *os, uint64_t object,
 	return (txh);
 }
 
+static dmu_tx_hold_t *
+dmu_tx_hold_object_impl(dmu_tx_t *tx, objset_t *os, uint64_t object,
+    enum dmu_tx_hold_type type, uint64_t arg1, uint64_t arg2)
+{
+	dnode_t *dn = NULL;
+	dmu_tx_hold_t *txh;
+	int err;
+
+	if (object != DMU_NEW_OBJECT) {
+		err = dnode_hold(os, object, tx, &dn);
+		if (err) {
+			tx->tx_err = err;
+			return (NULL);
+		}
+	}
+	txh = dmu_tx_hold_dnode_impl(tx, os, dn, type, arg1, arg2);
+	if (txh != NULL && txh->txh_dnode != NULL)
+		dnode_rele(txh->txh_dnode, tx);
```

Comment:
I don't understand the `if` statement here.  `dmu_tx_hold_dnode_impl()` adds its own hold to the dnode.  Therefore I think we should always release our hold, if we put one on.  We can also use `FTAG`, since this func is the only one adding/removing its hold on the dnode.  So this code should become:
```
if (object != DMU_NEW_OBJECT) {
    err = dnode_hold(os, object, FTAG, &dn);
    ...
...
if (dn != NULL) {
    dnode_rele(dn, FTAG);
}
```